apps/dashboard/lib/auth/workos.ts

```diff
@@ -0,0 +1,662 @@
+      orgs.map(org => [org.id, org.name])
+    );
+
+    return {
+      data: memberships.data.map((membership) => {
+      
+        return {
+          id: membership.id,
+          user,
+          organization: {
+            name: orgMap.get(membership.organizationId) || "Unknown Organization",
+            id: membership.organizationId
+          },
+          role: membership.role.slug,
+          createdAt: membership.createdAt,
+          updatedAt: membership.updatedAt,
+          status: membership.status
+        };
+      }),
+      metadata: memberships.listMetadata || {}
+    };
+  }
+  
+  async signUpViaEmail(params: { firstName: string, lastName: string, email: string }): Promise<void> {
+    const { email, firstName, lastName} = params;
+    if (!email || !firstName || !lastName) {
+      throw new Error(AuthErrorCode.MISSING_REQUIRED_FIELDS);
+    }
+  
+    try {
+      // create the user with the user details
+      // email will be unverified until they respond to the magic auth prompt
+      await WorkOSAuthProvider.provider.userManagement.createUser({
+        firstName,
+        lastName,
+        email
+      }); 
+  
+      await WorkOSAuthProvider.provider.userManagement.createMagicAuth({ email });
+    }
+    catch (error) {
+      const authError = error as AuthProviderError;
+      
+      // account exists
+      if (authError.errors?.some(detail => detail.code === 'email_not_available')) {
+        throw new Error(AuthErrorCode.EMAIL_ALREADY_EXISTS);
+      }
+
+      // malformed email
+      if (authError.message.includes("email_required")) {
+        throw new Error(AuthErrorCode.INVALID_EMAIL);
+      }
+  
+      // generic WorkOS error
+      if (authError.code === 'user_creation_error') {
+        throw new Error(AuthErrorCode.USER_CREATION_FAILED);
+      }
+  
+      // Generic error fallback
+      throw new Error(AuthErrorCode.UNKNOWN_ERROR);
+    }
+  }
+
+  async signIn(orgId?: string): Promise<void> {
+    throw new Error("Method not implemented.");
+  }
+
+  signInViaOAuth({ 
+    redirectUrl = env().NEXT_PUBLIC_WORKOS_REDIRECT_URI, 
+    provider,
+    redirectUrlComplete = SIGN_IN_REDIRECT
+  }: SignInViaOAuthOptions): string {
+    if (!provider) {
+      throw new Error('Provider is required');
+    }
+
+    const state = encodeURIComponent(JSON.stringify({ redirectUrlComplete }));
+
+    return WorkOSAuthProvider.provider.userManagement.getAuthorizationUrl({
+      clientId: WorkOSAuthProvider.clientId,
+      redirectUri: redirectUrl,
+      provider: provider === "github" ? "GitHubOAuth" : "GoogleOAuth",
+      state
+    });
+  }
+
+  async completeOAuthSignIn(callbackRequest: Request): Promise<OAuthResult> {
+    const url = new URL(callbackRequest.url);
+    const code = url.searchParams.get('code');
+    const state = url.searchParams.get('state');
+    const redirectUrlComplete = state 
+      ? JSON.parse(decodeURIComponent(state)).redirectUrlComplete 
```

Comment:
This could blow up if it's not valid json, I see below (line 335) we gracefully handle an error, should we do the same here?